File: cocos2dx_lua/VisualUI/ui/UILayer.lua
local UILayer=class("UILayer", function(data, parent)
    return cc.Node:create()
end)
cc.exports.UILayer=UILayer

function UILayer:ctor(data, parent)
    UIUtils.CocosGenBaseNodeByData(UIUtils.GetCurJsonData(data), self, true, self)
end

function UILayer:eventListener(event)
    local nodeName = event.target and (event.target:getName() or "")
    if string.len(nodeName or "") > 0 then
        local defaultName = "tap_" .. nodeName
        if not self[defaultName] then
            return
        end
        if tolua.type(event.target) == "ccui.Button" then
            if event.name == "ended" then
                self[defaultName](self, event)
            end
        else
            self[defaultName](self, event)
        end
    end
    dump(event)
end

File: cocos2dx_cpp/VisualUI/ui/UILayer.cpp
#include "UILayer.h"
#include <functional>
#include "../utils/UIUtils.h"

UILayer::UILayer(std::string data, cocos2d::Node* parent)
{
	UIUtils::CocosGenBaseNodeByData(UIUtils::GetCurJsonData(data), this, true, this);
}

UILayer* UILayer::create(std::string data, cocos2d::Node* parent)
{
	auto layer = new UILayer(data, parent);
	return layer;
}

void UILayer::eventListener(UIEvent& event)
{
	cocos2d::log("eventListener name is %s", event.name.c_str());
}

void UILayer::AddEventListener(std::string name, const eventCallback& callback)
{
	listeners[name] = callback;
}

void UILayer::RemoveEventListener(std::string name)
{
	auto iter = listeners.find(name);
	if (iter != listeners.end()) {
		listeners.erase(iter);
	}
}

const eventCallback UILayer::GetEventListener(std::string name)
{
	auto iter = listeners.find(name);
	if (iter != listeners.end()) {
		return iter->second;
	}
	
	return std::bind(&UILayer::eventListener, this, std::placeholders::_1);
}

